pkg/webhook/handlers.go

```diff
@@ -0,0 +1,165 @@
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/builder"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+)
+
+// Validator defines functions for validating an operation
+type Validator interface {
+	runtime.Object
+	ValidateCreate() error
+	ValidateUpdate(old runtime.Object) error
+}
+
+// Defaultor defines functions for setting defaults on resources
+type Defaulter interface {
+	runtime.Object
+	Default()
+}
+
+// NewResourceWebhook creates a new WebhookBuilder for Validation and Defaulting the provided type.  Returns nil
+// if o does not implement either Validator or Defaulter.
+func NewResourceWebhook(o runtime.Object, restMapper meta.RESTMapper, scheme *runtime.Scheme) (*admission.Webhook, error) {
+	// Create the handler
+	h := &resourceHandler{}
+	found := false
+	if v, ok := o.(Validator); ok {
+		h.Validator = v
+		found = true
+	}
+	if d, ok := o.(Defaulter); ok {
+		h.Defaulter = d
+		found = true
+	}
+	if !found {
+		// Type doesn't implement functions for webhooks
+		return nil, nil
+	}
+
+	// Create the webhook
+	return builder.NewWebhookBuilder().
+		Mutating(). // Always use a mutating webhook in case defaulting is implemented later
+		Operations(
+			admissionregistrationv1beta1.Create,
+			admissionregistrationv1beta1.Update).
+		WithRestMapper(restMapper).
+		WithScheme(scheme).
+		ForType(o).
+		Handlers(h).
+		Build()
+}
+
+type resourceHandler struct {
+	Validator Validator
+	Defaulter Defaulter
+	Decoder   types.Decoder
+}
+
+var _ inject.Decoder = &resourceHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *resourceHandler) InjectDecoder(d types.Decoder) error {
+	h.Decoder = d
+	return nil
+}
+
+// Handle handles admission requests.
+func (h *resourceHandler) Handle(ctx context.Context, req types.Request) types.Response {
+	// Get the object in the request
+	if resp, send := h.doValidate(ctx, req); send {
+		return resp
+	}
+
+	if resp, send := h.doDefault(ctx, req); send {
+		return resp
+	}
+
+	return admission.ValidationResponse(true, "")
+}
+
+func (h *resourceHandler) doDefault(ctx context.Context, req types.Request) (types.Response, bool) {
+	if h.Defaulter == nil {
+		return types.Response{}, false
+	}
+
+	// Get the object in the request
+	obj := h.Defaulter.DeepCopyObject().(Defaulter)
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err), true
+	}
+
+	// Default the object
+	copy := obj.DeepCopyObject().(Defaulter)
+	copy.Default()
+
+	// Create the patch
+	return admission.PatchResponse(obj, copy), true
```

Comment:
This causes roundtripping issue, we are moving away from this https://github.com/kubernetes-sigs/controller-runtime/pull/256 .